components/serializers/controller.cpp

```diff
@@ -0,0 +1,123 @@
+//#define SERIALIZER_DEBUG
+
+#ifdef SERIALIZER_DEBUG
+#include <iostream>
+#endif
+
+#include <boost/foreach.hpp>
+
+// Inconveniently, all of the controller classes are virtual because SceneUtil::Controller
+// is virtual, which expects the controller class to override apply() or otehr functions,
+// and as a result, we have to switch from static_cast to dynamic_cast in the Serializer
+// for these classes.  This situation has been anticipated by the OSG developers, because
+// they conveniently(?) provide a #define to change this behavior.
+#undef OBJECT_CAST
+#define OBJECT_CAST dynamic_cast
+
+#include <osgDB/ObjectWrapper>
+#include <osgDB/InputStream>
+#include <osgDB/OutputStream>
+#include <osgDB/Serializer>
+#include "fixes.hpp"
+
+#include <components/nifosg/nifloader.hpp>
+
+#include "keymap.hpp"
+
+SERIALIZER_KEYMAPT(XRotations, NifOsg::FloatInterpolator::InnerMapType, float, NifOsg::KeyframeController)
```

Comment:
I really don't care for the macro programming approach here, but that's how OSG does it, and I wasn't inclined to rock the boat too much here.   I think if I were rewriting the serializer from scratch, I would have tried to do it all with templates, but maybe there's some non-obvious problem why that doesn't work.   The author certainly seems to have known how to use templates.